[c2] framework for committed serialized tests

ajyu opened the pull request:
Summary:
Generate serialized test inputs/outputs/backward graphs of tests inside `caffe2/python/operator_test` that call assertSerializedOperatorCheck(). Tests should be decorated with serialized_test.collect_tests.given_and_seeded to run hypothesis tests that are actually random and a single fixed seeded hypothesis tests.

To use:
1. Refactor your test to be a SerializedTestCase
1a. Decorate it with @given_and_seeded
1b. Call testWithArgs in main
2. Run your test with -g to generate the output. Check it in.
3. Subsequent runs of the test without generating the output will check against the checked in test case. 

Details:
Run your test with `python caffe2/python/operator_test/[your_test].py -g`
Outputs are in `caffe2/python/serialized_test/data`. The operator tests outputs are in a further subdirectory `operator_test`, to allow for other tests in the future (model zoo tests?)

Currently, we've only refactored weighted_sum_test to use this, but in the next diff, we'll refactor as many as possible. The directory structure may also change as usually there are multiple tests in a single file, so we may create more structure to account for that.

Test Plan:
1. Run `python caffe2/python/operator_test/weighted_sum_test.py -g` to generate the outputs
2. Run `python caffe2/python/operator_test/weighted_sum_test.py`; Check that the unit test itself still runs as expected
3. Run `python caffe2/python/gradient_check_test.py`; Check that CheckSimple still works as expected
4. Run `with-proxy FULL_CAFFE2=1 python setup.py develop`; Check there are no errors

ajyu commented on caffe2/python/operator_test/weighted_sum_test.py:
```diff
@@ -52,10 +52,18 @@ def weighted_sum_op_ref(*args):
                 outputs_with_grads=[0],
             )
+
+        if should_register:
```
Would like your thoughts about wether this pattern is ok.

houseroad commented on caffe2/python/serialized_test/generate_serialized_tests.py:
```diff
@@ -0,0 +1,51 @@
+DATA_DIR = os.path.join(TOP_DIR, 'data')
+
+
+def serialize_tests(args):  # type: (argparse.Namespace) -> None
```
We probably don't need this mypy annotation. But having them won't hurt.

houseroad commented on caffe2/python/serialized_test/generate_serialized_tests.py:
```diff
@@ -0,0 +1,51 @@
+
+import caffe2.python.serialized_test.collect_tests as collect
+from typing import Text
+import pickle
```
Please don't use pickle. It's not safe, and if we update python classes, we may not be able to load the test data.
Please serialize the operator through protobuf, and data serialized using .npz format.

Here is some example: 
https://github.com/pytorch/pytorch/blob/master/test/onnx/test_operators.py#L74

houseroad commented on caffe2/python/serialized_test/serialized_test_data.py:
```diff
@@ -0,0 +1,13 @@
+
+from collections import namedtuple
+
+SerializedTestData = namedtuple('SerializedTestData', [
```
I guess we probably don't need this, since we can use the folder to enocde test type and anme.
Operators can be serilized as op.pb, and you probably want to support multiple sets of inputs and outputs.

houseroad commented on caffe2/python/operator_test/weighted_sum_test.py:
```diff
@@ -4,20 +4,21 @@
-    @given(n=st.integers(5, 8), m=st.integers(1, 1),
-           d=st.integers(2, 4), grad_on_w=st.booleans(),
-           **hu.gcs_cpu_only)
+    @serial.given_and_seeded(
```
Can we don't change given, but have another separate wrapper, called `serialize_with_seed()`

ajyu replied:
We had previously tried to do this, but it proved to be tricky because of the following
1. hypothesis.given annotates a test with a bunch of variables
https://hypothesis.readthedocs.io/en/latest/_modules/hypothesis/core.html
2. when we define a decorator outside of hypothesis, we don't have access to those variables for modification, because we don't assume we know anything about the wrapped function

houseroad commented on caffe2/python/operator_test/weighted_sum_test.py:
```diff
@@ -4,20 +4,21 @@
     def test_weighted_sum(self, n, m, d, grad_on_w, gc, dc):
         input_names = []
         input_vars = []
+        np.random.seed(2603)
```
Can we add the seed as a parameter in serialize_with_seed (the annotation?)

ajyu replied:
Sure thing

houseroad commented on caffe2/python/operator_test/weighted_sum_test.py:
```diff
@@ -52,10 +53,16 @@ def weighted_sum_op_ref(*args):
                 outputs_with_grads=[0],
             )
+
+        self.assertSerializedOperatorCheck(input_vars, outs, grad_op)
```
Shall we just hide this call into assertGradientChecks?

houseroad commented:
CI shows the binary data cannot be found

ajyu commented:
Yeah, that's unfortunate. I'll probably have to look into the CI job to see what command they're using to build

houseroad commented on caffe2/python/hypothesis_test_util.py:
```diff
@@ -318,6 +318,40 @@ def runOpBenchmark(
+            )
+        net = core.Net("opnet")
+        net.Proto().op.extend([op])
+        workspace.RunNetOnce(net)
```
Nit: Since input is OperatorDef, use RunOperatorOnce is more natural. 

Example is here:
https://github.com/pytorch/pytorch/blob/342517e6e72f51048a4b497ba700ecd651bd26ab/caffe2/python/onnx/backend.py#L245

ajyu replied:
Sure thing, will change.

houseroad commented on setup.py:
```diff
@@ -1068,6 +1068,23 @@ def make_relative_rpath(path):
     }
 
+
+def collect_data_files(directory):
```
Since we modify the CMake file, is this part of the change is needed?

ajyu replied:
I think it would be necessary if this were to be run from setup.py. It's here that we choose between cmake and setup.py. Even though it's not currently an "integrated" test, I guess it could be in the future. I can take this out or leave it in, your choice.
https://github.com/pytorch/pytorch/blob/master/.jenkins/caffe2/build.sh#L199

houseroad replied:
The question is that if we install the caffe2 using `FULL_CAFFE2=1 python setup.py install`, is the change on CMakeLists is already enough or we still need the changes in setup.py. Could you verify this and make sure we don't bring in any unnecessary change?

ajyu replied:
I manually verified that if we do not modify setup.py, that the expected binary files don't show up in the output path used by setuptools, `/home/ansha/venv-install/lib64/python2.7/site-packages/caffe2/python/serialized_test/data`

houseroad replied:
ok, then let's keep this change.

houseroad commented on caffe2/python/serialized_test/serialized_test_util.py:
```diff
@@ -0,0 +1,192 @@
+            device_option,
+        )
+
+
```
Nit: could you add the usage of the script in the comments? Or add `README.md` in `caffe2/python/serialized_test/`.

houseroad reviewed (approved):
Overall, looks good. Some nits to address. :-)

houseroad commented on caffe2/python/serialized_test/serialized_test_util.py:
```diff
@@ -0,0 +1,192 @@
+        numpy.savez_compressed(
+            os.path.join(output_dir, 'outputs'), outputs=outputs)
+
+    def compare_test(self, inputs, outputs, grad_ops):
```
Nit: let's support precision/tolerance here

ajyu commented:
@pytorchbot retest this please

ezyang commented:
Can you explain why setup.py needs to register these serialized tests as `data_files`? I don't think we want to distribute them as part of the package.

ajyu commented:
I needed to add the files to CMakeLists in order for the files to be included in the Jenkins test build to pass tests. I figured that making the setup.py build work as well would be desired, in case we switch Jenkins to use it rather than cmake. But I can take that out if desired.

ezyang reviewed (changes requested):
OK, in that case, let's remove it from `data_files`. You should figure out another way to get Jenkins tests to pass. Is the problem that these files are generated at build time and so they get lost when we move files from build to test time? There's some prior art about this in `.jenkins/pytorch/build.sh`, see:

```

# Add the test binaries so that they won't be git clean'ed away
git add -f build/bin
```

ajyu commented:
The files are not generated at build time, they live in the codebase and engineers would generate them at the time they write their tests. They are copied during build time, and I guess subsequently copied during test time.

I'll remove the changes from setup.py then.

Are repo files available to Jenkins during the time of testing? If so, I can try to figure out to directly read from the repo location if so. Otherwise, I'd need to keep the CMakeLists.txt changes to allow Jenkins to continue working.

ezyang commented:
In that case, yes, all source controlled files are available during tests, at the source code locations.

By the way, how big are these files going to be? If someone checks in a large binary file, that binary size is *permanent* in the git repository, and constitutes a tax on everyone who every git clones the repo at any point in the future from there. Let's make it hard for people to accidentally do this.

houseroad commented:
@ezyang for the model and input/output data, they are small, just like the test data we committed to onnx repo. They should be fine. The goal to detect the incompatible changes on gradient ops to prevent sev (live the broadcasting one).

orionr reviewed (changes requested):
Great work! I'm curious what the size of the test files is, but if small we're in good shape if you can remove the CMakeLists.txt changes. If this requires some CI changes as well, let me know and we can adjust them.

orionr commented on caffe2/python/operator_test/weighted_sum_test.py:
```diff
@@ -4,20 +4,22 @@
 import caffe2.python.hypothesis_test_util as hu
 import hypothesis.strategies as st
 import numpy as np
+import caffe2.python.serialized_test.serialized_test_util as serial
```
nit: Could move this up after `caffe2.python.hypothesis_test_util` to be sorted.

orionr reviewed (approved):
Looks good!

ajyu committed:
framework for committed serialized tests (#10594)

Summary:
Generate serialized test inputs/outputs/backward graphs of tests inside `caffe2/python/operator_test` that call assertSerializedOperatorCheck(). Tests should be decorated with serialized_test.collect_tests.given_and_seeded to run hypothesis tests that are actually random and a single fixed seeded hypothesis tests.

To use:
1. Refactor your test to be a SerializedTestCase
1a. Decorate it with given_and_seeded
1b. Call testWithArgs in main
2. Run your test with -g to generate the output. Check it in.
3. Subsequent runs of the test without generating the output will check against the checked in test case.

Details:
Run your test with `python caffe2/python/operator_test/[your_test].py -g`
Outputs are in `caffe2/python/serialized_test/data`. The operator tests outputs are in a further subdirectory `operator_test`, to allow for other tests in the future (model zoo tests?)

Currently, we've only refactored weighted_sum_test to use this, but in the next diff, we'll refactor as many as possible. The directory structure may also change as usually there are multiple tests in a single file, so we may create more structure to account for that.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/10594

Differential Revision: D9370359

fbshipit-source-id: 8688cfca5daa46adcfec1e872a12e4342145af88

ezyang reviewed (approved):
I didn't review the non-setup/jenkins bits of the patch, but those parts now look extremely good now. Thanks!

apaszke commented:
Do we really believe that checked in tests are a good idea? I'm pretty sure that having more rigorous test harness will provide much better coverage, and have a benefit of not checking in a new file after every single operator change. Here's how I see it:

Operator code is generally quite self-contained, so it's unlikely that a change in one part of the system will affect a random operator on the other end. Unless it's a breakage so bad that most operator tests will be red anyway. On the other hand, if someone changes the operator kernels, the tiny numerical differences might change e.g. the few least-significant bits, raising a failure when compared to the expected test. Since this was a PR that updates an op, I believe the author will be more than happy to check in a new set of expected outputs, meaning that those values don't mean much. Finally, you can't test all ops this way, because if you use an external library, just updating it to a new version (or having a different cuDNN version installed) might result in those tiny differences, flaring up the whole test suite for no good reason.

We actually took a similar approach for testing the PyTorch JIT some time ago (we serialize IR as text, not binary tensor values), but it has many scalability problems, and is notoriously hard to review. We already had many situations where the test had a good expect file checked in, only to have it overwritten with a non-working one in a later PR (and we do pretty careful code review). After the release, we'll probably try to move away from comparing to expected values, and you might want to consider that too.
